File: finanze/domain/external_entity.py
from datetime import datetime
from enum import Enum
from typing import Any, Optional
from uuid import UUID

from domain.entity import Entity, EntityType, Feature
from domain.external_integration import (
    ExternalIntegrationId,
)
from pydantic.dataclasses import dataclass


class ExternalEntityStatus(str, Enum):
    UNLINKED = "UNLINKED"
    LINKED = "LINKED"
    ORPHAN = "ORPHAN"


@dataclass
class ExternalEntity:
    id: UUID
    entity_id: UUID
    status: ExternalEntityStatus
    provider: ExternalIntegrationId
    date: Optional[datetime] = None
    provider_instance_id: Optional[str] = None
    payload: Optional[dict] = None


EXTERNAL_ENTITY_FEATURES = [Feature.POSITION]


@dataclass
class ProviderExternalEntityDetails:
    id: str
    name: str
    bic: str
    type: EntityType
    icon: Optional[str]


@dataclass
class ExternalEntityCandidates:
    entities: list[ProviderExternalEntityDetails]


@dataclass
class ExternalEntityCandidatesQuery:
    providers: Optional[list[ExternalIntegrationId]]
    country: Optional[str]


@dataclass
class ExternalEntityLoginRequest:
    external_entity: ExternalEntity
    redirect_host: Optional[str] = None
    relink: bool = False
    institution_id: Optional[str] = None
    user_language: Optional[str] = None


@dataclass
class ExternalEntityFetchRequest:
    external_entity: ExternalEntity
    entity: Entity


class ExternalEntitySetupResponseCode(str, Enum):
    ALREADY_LINKED = "ALREADY_LINKED"
    CONTINUE_WITH_LINK = "CONTINUE_WITH_LINK"


@dataclass
class ExternalEntityConnectionResult:
    code: ExternalEntitySetupResponseCode
    link: Optional[str] = None
    provider_instance_id: Optional[str] = None
    payload: Optional[Any] = None
    id: Optional[UUID] = None


@dataclass
class ExternalFetchRequest:
    external_entity_id: UUID


@dataclass
class ConnectExternalEntityRequest:
    institution_id: Optional[str]
    external_entity_id: Optional[UUID]
    provider: Optional[ExternalIntegrationId]
    relink: bool = False
    redirect_host: Optional[str] = None
    user_language: Optional[str] = None


@dataclass
class CompleteExternalEntityLinkRequest:
    payload: Optional[dict] = None
    external_entity_id: Optional[str] = None


@dataclass
class DeleteExternalEntityRequest:
    external_entity_id: UUID

File: mobile/src/domain/core/externalEntity.ts
import { Entity, EntityType } from "./entity"
import { ExternalIntegrationId } from "./externalIntegration"

export enum ExternalEntityStatus {
  UNLINKED = "UNLINKED",
  LINKED = "LINKED",
  ORPHAN = "ORPHAN",
}

export interface ExternalEntity {
  id: string
  entityId: string
  status: ExternalEntityStatus
  provider: ExternalIntegrationId
  date?: string | null
  providerInstanceId?: string | null
  payload?: Record<string, any> | null
}

export interface ProviderExternalEntityDetails {
  id: string
  name: string
  bic: string
  type: EntityType
  icon: string | null
}

export interface ExternalEntityCandidates {
  entities: ProviderExternalEntityDetails[]
}

export interface ExternalEntityCandidatesQuery {
  providers: ExternalIntegrationId[] | null
  country: string | null
}

export interface ExternalEntityLoginRequest {
  externalEntity: ExternalEntity
  redirectHost?: string | null
  relink?: boolean
  institutionId?: string | null
  userLanguage?: string | null
}

export interface ExternalEntityFetchRequest {
  externalEntity: ExternalEntity
  entity: Entity
}

export enum ExternalEntitySetupResponseCode {
  ALREADY_LINKED = "ALREADY_LINKED",
  CONTINUE_WITH_LINK = "CONTINUE_WITH_LINK",
}

export interface ExternalEntityConnectionResult {
  code: ExternalEntitySetupResponseCode
  link?: string | null
  providerInstanceId?: string | null
  payload?: any | null
  id?: string | null
}

export interface ExternalFetchRequest {
  externalEntityId: string
}

export interface ConnectExternalEntityRequest {
  institutionId: string | null
  externalEntityId: string | null
  provider: ExternalIntegrationId | null
  relink?: boolean
  redirectHost?: string | null
  userLanguage?: string | null
}

export interface CompleteExternalEntityLinkRequest {
  payload?: Record<string, any> | null
  externalEntityId?: string | null
}

export interface DeleteExternalEntityRequest {
  externalEntityId: string
}
